crates/error/src/error.rs

```diff
@@ -0,0 +1,1751 @@
+                for e in Chain::new(e) {
+                    write!(f, ": {e}")?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<E> From<E> for Error
+where
+    E: core::error::Error + Send + Sync + 'static,
+{
+    fn from(error: E) -> Self {
+        Self::new(error)
+    }
+}
+
+impl From<Error> for Box<dyn core::error::Error + Send + Sync + 'static> {
+    #[inline]
+    fn from(error: Error) -> Self {
+        error.into_boxed_dyn_error()
+    }
+}
+
+/// Convert a [`Error`] into an [`anyhow::Error`].
+///
+/// # Example
+///
+/// ```
+/// # use wasmtime_internal_error as wasmtime;
+/// let wasmtime_error = wasmtime::Error::msg("whoops");
+/// let anyhow_error = anyhow::Error::from(wasmtime_error);
+/// ```
+//
+// Unfortunately, we can't also implement `From<anyhow::Error> for Error`
+// because of trait coherence. From Rust's trait system's point of view,
+// `anyhow` could theoretically add an `core::error::Error for anyhow::Error`
+// implementation, which would make our desired `From<anyhow::Error>`
+// implementation conflict with our existing `From<E: core::error::Error>`
+// implementation. They cannot in fact add that implementation, however, because
+// they already have a `From<E: core::error::Error> for anyhow::Error`
+// implementation and so adding `core::error::Error for anyhow::Error` would
+// cause that impl to conflict with `From<T> for T` (which is the same reason we
+// cannot implement `core::error::Error for Error`). Nonetheless, our hands are
+// tied here.
+#[cfg(feature = "anyhow")]
+impl From<Error> for anyhow::Error {
+    #[inline]
+    fn from(e: Error) -> Self {
+        anyhow::Error::from_boxed(e.into_boxed_dyn_error())
+    }
+}
+
+impl Error {
+    /// Construct a new `Error` from a type that implements
+    /// `core::error::Error`.
+    ///
+    /// Calling [`error.is::<E>()`][Error::is] will return `true` for the new
+    /// error (unless there was a memory allocation failure).
+    ///
+    /// This boxes the inner error, but if that box allocation fails, then this
+    /// function returns an `Error` where
+    /// [`error.is::<OutOfMemory>()`][crate::OutOfMemory] is true.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmtime_internal_error as wasmtime;
+    /// use wasmtime::Error;
+    ///
+    /// let error = Error::new(std::fmt::Error);
+    /// ```
+    pub fn new<E>(error: E) -> Self
+    where
+        E: core::error::Error + Send + Sync + 'static,
+    {
+        if TypeId::of::<E>() == TypeId::of::<OutOfMemory>() {
+            return Error {
+                inner: OutOfMemory::new().into(),
+            };
+        }
+
+        return Self::from_error_ext(ForeignError(error));
+
+        // NB: The `repr(transparent)` is required for safety of the
+        // `ErrorExt::ext_is` implementation and the casts that are performed
+        // using that method's return value.
+        #[repr(transparent)]
+        struct ForeignError<E>(E);
```

Comment:
Is there a reason this has to be nested within `new()`? I understand the encapsulation is nice (it's only instantiated here, nowhere else) but the indents are a little deep and it makes the code harder to read for me at least -- in-place struct impls are nicest when they're small or ad-hoc IMHO.

(Not a huge deal, just my aesthetic preference)

Reply:
No reason besides making it clear what the encapsulation is, but I can change it.